### DISCUSSION TITLE
[WASM] Is Uno WASM a viable option for LineOfBusiness (LOB) applications?

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#7949] [General]
Hi,

I ran some tests using the [UNO Gallery ](https://gallery.platform.uno/)app. 

Here are my findings (latest Chrome on latest Windows 10 using a 1GB line):
- Chrome console says "~26MB transferred" on each start
- startup takes 8-10 secs until the first content is visualized
- forcing a fresh reload (CTRL+F5) and restart does not change much on these numbers
- VS 2022->Analyze->Calculate Code Metrics on Uno.Gallery solution (I deleted all heads, only left the WASM head) says ~1230 lines of executable code 

I'm trying to get a grasp on what this means for a non-trivial LOB app (e.g. my app has > 30K lines of executable code and still is not large by any stretch)

- why does UNO WASM not benefit from the browser cache (at all? noticeably?), neither in amount of data transferred nor in startup time?
- even the (relatively small?) Uno Gallery app has a significant startup time which is way longer than what the average user would expect/tolerate (2-3 secs)
- considering any "real world LOP app" I'm worried the numbers which had to be expected there

What's UNO's stance on these findings? What could be done to improve the startup experience?

Thanks in advance for shedding some light

**Comment by jeromelaban** [comment]
We've been working in performance for the upcoming Uno 4.1 update, and the gallery application will benefit from these changes. 

This particular app is a showcase which is currently not using Profile guided AOT (hence the 26MB), is using both Material and Cupertino styles (it's generally not the case in normal apps), and is not yet updated to Uno 4.1. Updates will be coming soon.

Finally, .NET 6 still has issues (https://github.com/dotnet/runtime/issues/60247) with AOT and interpreter/AOT boundaries which are being worked on for .NET 7. We expect that when these paths are optimized, runtime performance will increase significantly.

Each app has its own performance characteristics, I suggest to try out uno with your code to find out which and let us know.

**Reply by DierkDroth** [reply]
Thanks for your feedback @jeromelaban. I'm definitely looking forward 4.1

On .NET 7: Does your comment imply, that you would believe that performance/download size issues only would be fixed with .NET7 (end of this year) and not earlier?

Lastly: as it stands I will be using Uno's Material approach. That's what made me looking into the gallery app and trying to understand what I'm buying into...

**Reply by DierkDroth** [reply]
@jeromelaban would you have any insight why - despite large (initial) download size - the gallery app is no benefiting from the browser cache? I would believe this (mis-?) behavior would be expected for any UNO WASM app then, no?

I guess that's the part which concerns me most - initial download size for most LOB apps likely is secondary. What's important is that - after having 'installed' the app - it should start 'quick as possible' and not go through a (perceived? - but definitely measurable) download cycle again

**Reply by jeromelaban** [reply]
Uno is going to be using .NET 7 long before it's released officially; performance improvements are going to be available sooner.

For the wasm caching, browsers have a maximum size for wasm payload w.r.t caching. If it gets too big, the compilation cache is not used. Download cache can also be affected by this, though the threshold is likely different per browser and versions.

**Reply by DierkDroth** [reply]
Thanks for your feedback @jeromelaban

**Reply by DierkDroth** [reply]
@jeromelaban I just ran another test with UNO Gallery on Chrome with F12/console up and saw that only ~6.5MB are transferred. This would be a **significant improvement over my last test**.

- is UNO gallery updated to UNO 4.1 yet?
- is the improvement I observed factual or am I ´chasing shadows´ or comparin apples to oranges here?

**Reply by jeromelaban** [reply]
The application is using [PG-AOT](https://github.com/unoplatform/Uno.Wasm.Bootstrap#profile-guided-aot), which is why the size has gone down significantly.

**Reply by DierkDroth** [reply]
Understand. Thanks for clarification @jeromelaban

**Reply by DierkDroth** [reply]
OK, I finally got around building a RELEASE version (using latest UNO 4.2 bits) of my app + uploading it and found that a hefty ~85 MB are downloaded.

The massive download size makes me wonder if there is compression effective at all? Note: I did not touch anything in that regard, but just uploaded the complete ´publish´ build folder + started the app from the ´index.html´ file with in that folder.

How could the download size be reduced substantially?

<img width="1280" alt="Untitled" src="https://user-images.githubusercontent.com/26179242/158049985-be7647d6-7646-444a-9788-9309ff7a0112.png">

**Reply by jeromelaban** [reply]
Please see the PG-AOT documentation mentioned here: https://github.com/unoplatform/uno/discussions/7949#discussioncomment-2339040

**Reply by DierkDroth** [reply]
@jeromelaban yes I´m aware of PG AOT. However, I´m pretty sure there is something wrong with compression in my case above - I suspect that simply the uncompressed version was downloaded. 

How could I check and make sure the compressed version really is used (preferrably even the most efficient Brotli compression)?

**Reply by jeromelaban** [reply]
It all depends on how you're hosting your application. You can check the response headers to find out if compression is enabled or not. You can compare the headers with https://gallery.platform.uno.

**Reply by DierkDroth** [reply]
@jeromelaban you´ll see me scratching my head as I´m trying to understand what you´re telling me :-)

I have no idea what response headers are, nor what´s the relevance on whether or not compression is used with UNO WASM nor how I could configure things either way. Would you have some pointers so I could educate myself on how response headers are relevant in the UNO WASM case and how I could configure/test things?

So far I simply uploaded the complete ´publish´ folder and started the browser with the correct URL. That´s it. Appears there´s much more to worry about?!?

Thanks in advance

**Reply by DierkDroth** [reply]
@jeromelaban alright I did a laymen´s test and compared the sizes of dotnet.wasm as downloaded by Uno.Gallery vs. in my app.

This is 4.8MB vs. ~86MB!

This is a prove that my app is not using compression, correct?
If so: how could I make sure compression is used as the browser downloads my app?

**Reply by jeromelaban** [reply]
By default Uno's output [pre-compression](https://github.com/unoplatform/Uno.Wasm.Bootstrap#pre-compression-support) only supports IIS (using web.config) and azure static apps (using a .br extension) for compression. If you use something else, you'll have to configure compression yourself.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban I now have something to work with.

However, as I was reading through the [link you provided above](https://github.com/unoplatform/Uno.Wasm.Bootstrap#pre-compression-support) I found that the link "the following file" throws a 404 error. 

I believe that file would be crucial for me efforts, no? Could that link please be fixed? Thanks

**Reply by jeromelaban** [reply]
Any web.config file from the repo can do, like [this one](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/src/Uno.Wasm.AotTests/wwwroot/web.config).

**Reply by DierkDroth** [reply]
@jeromelaban there already was a web.config file in the wwwroot subdirectory which made it to ´publish/wwwroot/web.config´ and ´publish/web.config´. However I replaced the file with the version which you provided above (which is different). After upload it still shows that dotnet.wasm would have a size of 86MB.

This is an AWS S3 bucket. 

Any idea what´s going on?

**Reply by jeromelaban** [reply]
Uno supports IIS and Azure Static Apps by default. Other platforms/services need to [be configured manually](https://platform.uno/docs/articles/how-to-host-a-webassembly-app.html?q=server). If you get your app to be served with compression, we'd like to have your contributions to the documentation.

**Reply by DierkDroth** [reply]
Here are my latest findings:

@9horses - my cloud admin - established compression on my AWS S3 space (@9horses announced to provide tech details later). This brought down the download size from ~86MB to ~23MB using latest **official** UNO bits (no pre-release nor any PGAOT involved a this point).
Additional benefit: when going to the same link again all content is loaded from browser cache - no network traffic involved (latest Chrome on Windows 11)

This is a significant improvement!

However, the download size still is massive and I wonder if upcoming MS or UNO updates/improcements would promise even better results (PGAOT aside). @jeromelaban would you have any insights to share?

**Reply by jeromelaban** [reply]
We do not know the plans for Microsoft on that topic. On Uno's side, [PG-AOT ](https://github.com/unoplatform/Uno.Wasm.Bootstrap#profile-guided-aot), the [IL Linker](https://platform.uno/docs/articles/features/using-il-linker-webassembly.html?q=linker) and [Xaml Resource trimming](https://platform.uno/docs/articles/features/resources-trimming.html?q=trim) are the current features to reduce the payload size.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. 

My RELEASE configuration looks like this:
```  
<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <PlatformTarget>x86</PlatformTarget>
    <WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
</PropertyGroup>
```

Does this mean the IL Linker is ON so full tree shaking is in effect already?

**Reply by DierkDroth** [reply]
@jeromelaban 2 additional observations as of my latest tests:
a) adding `<UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>` to the 'PropertyGroup' above does exactly nothing. The download size as per Chrome 'Network' tab does not show the slightest difference. Is that expected?
b) The initial download after a fresh upload shows the ~23MB from above. If I then did CTRL+F5 in Chrome then it would download ~15MB. I thought CTRL+F5 would force a complete reload?!? Would you know what the difference is?

**Reply by jeromelaban** [reply]
`UnoXamlResourcesTrimming` works by configuring the linker, and if you changed the linker settings, this could effectively do nothing. Make sure to adjust what you're linking out.

Initial download behavior is very dependent on browser cache behavior, it's not something uno has control over.

**Reply by DierkDroth** [reply]
> UnoXamlResourcesTrimming works by configuring the linker, and if you changed the linker settings, this could effectively do nothing. Make sure to adjust what you're linking out.

I'm sorry but I'm not following. 
Where and how would I need to apply `<UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>` to make it effective for a RELEASE build (I don't want to have it effective for the DEBUG build to not slow down the DEBUG build)? To a different 'PropertyGroup'? Else?

**Reply by jeromelaban** [reply]
> UnoXamlResourcesTrimming works by configuring the linker, and if you changed the linker settings, this could effectively do nothing. 

Take a look at your linker settings, particularly if you have `uno.ui` in the `linkerconfig.xml` file.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban, I'm trying to understand...

I have this this in my linkerConfig.xml
```<linker>
  <assembly fullname="Unusual.Common" />
  <assembly fullname="Unusual.Wasm" />
  <assembly fullname="Uno.UI" />

  <assembly fullname="mscorlib" />
  <assembly fullname="System.Core" />
</linker>
```

Does this mean any assembly listed there is *excluded* from IL Linker tree shaking as well as *excluded* from any XAML resource trimming?

Also: is linkerConfig.xml effective for RELEASE build only or for DEBUG build as well?

**Reply by jeromelaban** [reply]
Yes, if you exclude Uno.UI from the linker, XAML resource trimming is not going to work. In general, you should not put Uno.UI there. Linker being enabled in release or debug is your choice in the csproj.

**Reply by DierkDroth** [reply]
@jeromelaban pardon my ignorance, but I'm trying to get 100% clarity on how this works.

I'm aware [of this](https://github.com/dotnet/linker/blob/main/docs/data-formats.md#xml-examples). So basically what I have in my linkerConfig.xml means:
- all assemblies listed as per above in my linkerConfig.xml are in 'preserve=all' mode
- this mean they would not touched by any tree-shaking logic at all
- they also would not be touched by any XAML resource trimming

Is that correct?

Next question then: why would I want to have *any file* in linkerConfig.xml in the first place? Wouldn't it rather make sense to start 'with nothing' and only add assemblies in case I ran into trouble? Or do we know already that some assemblies 'no matter what' should not be touched by tree-shaking nor XAML trimming?

**Reply by jeromelaban** [reply]
Assemblies placed in the linker config are force to be kept entirely in the resulting payload, including XAML resource trimming.

You'll want assemblies in the linker config to avoid reflection related issues, generally by the method specifically, but we do not give guidelines for that as it's up to the library authors to handle this.

As I mentioned previously, you do not need to place any of Uno's assemblies there, those are linker-compatible.

**Reply by DierkDroth** [reply]
Thanks for your patience @jeromelaban 

A few comments:
- I recommend adding some explicit documentation on what to do (or not) to configure the linker. This definitely could help the less experienced users like myself who have little understanding what's going on behind the scenes when building UNO WASM and don't know which setting is where and impacts what
- I just tested latest templates: linkerConfig.xml holds this line `<assembly fullname="Uno.UI" />` which is - according your comments above - not needed (please correct me if I'm wrong). I recommend removing that line from the templates.
- I would add the `preserve="all"` parameter to the linkerConfig.xml template, since it's more verbose than omitting it and makes it clearer what's going on

**Reply by DierkDroth** [reply]
Latest findings after removing UNO.UI from linkerConfig.xml + adding `<UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>` to the section below (project file):
- initial download: 22.1 MB
- CTRL+F5: 14.5 MB

-> only small improvement. Appears tree shaking and resource trimming does not do much to UNO.UI

```
<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
         <PlatformTarget>x86</PlatformTarget>
	 <UnoXamlResourcesTrimming>true</UnoXamlResourcesTrimming>
	 <WasmShellMonoRuntimeExecutionMode>InterpreterAndAOT</WasmShellMonoRuntimeExecutionMode>
</PropertyGroup>
```

**Comment by DierkDroth** [comment]
FYI summary of what I have so far:
- latest UNO packages as per screenshot below
- using an AWS S3 bucket as web space
- started uncompressed with ~86MB download size
- download size now is down to ~22MB due to:
-- Brotli compression
-- relatively small gains (details above) by [IL Linker](https://platform.uno/docs/articles/features/using-il-linker-webassembly.html?q=linker) and [XAML resource trimming](https://platform.uno/docs/articles/features/resources-trimming.html?q=trim)

I'm shying away from [PG-AOT](https://github.com/unoplatform/Uno.Wasm.Bootstrap#profile-guided-aot) for now, since I do have performance concern. However, I'll look into again as performance improves on [our way to .NET 7](https://github.com/unoplatform/uno/discussions/8417)

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/26179242/161292145-4e63a311-4363-4f69-8dff-1cd85a394950.png">
